Clarify documentation on multi-module projects (#9452)
quarkusio/quarkus: docs/src/main/asciidoc/getting-started.adoc

The test port can be controlled via the `quarkus.http.test-port` config property. Quarkus also creates a system
property called `test.url` that is set to the base test URL for situations where you cannot use injection.

== Working with multi-module project or external modules

> Quarkus heavily utilizes Jandex at build time, to discover various classes or annotations. One immediately recongizable of this, is CDI bean discovery.

Reviewer: ```suggestion
Quarkus heavily utilizes https://github.com/wildfly/jandex[Jandex] at build time, to discover various classes or annotations. One immediately recongizable application of this, is CDI bean discovery.
```

> Quarkus uses Jandex to create, at build time, an index of all classes eligible to CDI bean discovery.
> Without it, none of the Quarkus extensions will work.

Reviewer: ```suggestion
Quarkus heavily utilizes Jandex at build time, to discover various classes or annotations. One immediately recongizable of this, is CDI bean discovery.
As a result, most of the Quarkus extensions will work not work properly if this build time discovery isn't properly setup.
```

> This index is created by default on your project thanks to our Maven or Gradle plugin.

Reviewer 2: I am not sure that's accurate. Maven/gradle indexing plugins are configured by users in their projects if they choose to do that. We are indexing certain dependencies at build time (unless the index is already available).

Author: We automatically index _application classes_, that wasI refers to. If it's not via the Maven/Gradle plugin I can remove this part of the sentence.
But this means that you have nothin to do on your Quarkus projet ...

Reviewer 2: Ok, I probably misunderstood what you wanted to say here. My point was that it was not the Maven/plugins themselves that do that but given that they do launch the Quarkus build process which also includes the indexing phase, that's probably ok.

Author: Maybe we can find a better way to express this. It is Quarkus that creates the index on the project that bootstrap it, difficult to express this in a way that everybody understood.

Reviewer 2: Ok, how about this:

Quarkus uses Jandex to create, at build time, an index of all the classes eligible to CDI bean discovery. The application project classes will be indexed by Quarkus by default.

> This index is created by default on your project thanks to our Maven or Gradle plugin.

Reviewer: ```suggestion
This index is created by default on the project on which Quarkus is configured for, thanks to our Maven and Gradle plugins.
```

> As a result, most of the Quarkus extensions will work not work properly if this build time discovery isn't properly setup.

Reviewer: ```suggestion
As a result, most of the Quarkus extensions will not work properly if this build time discovery isn't properly setup.
```

> If your project is a multi-module project, be sure to read the Working with multi-module projects section on our

Reviewer: ```suggestion
However, when working with a multi-module project, be sure to read the `Working with multi-module projects` section of the
```

link:maven-tooling#multi-module-maven[Maven] or link:gradle-tooling#multi-module-maven[Gradle] guides.

If you plan to use external modules (for example, an external library for all your domain objects),

> you will need to setup the creation of the index for those modules via the Jandex plugin (if you can modify them)

Reviewer: ```suggestion
you will need to make these modules known to the indexing process either by adding the Jandex plugin (if you can modify them)
```

> or via the `quarkus.index-dependency` inside your `application.properties`.

Reviewer 2: Or `META-INF/beans.xml`.

Author: I wanted to be more directive than offering the three ways.

The purpose of the guide is to be straightfoward: _if you can add the maven plugin, add it, otherwise use the quarkus.index-dependency properties._

Reviewer 2: Sure. My personal view is `META-INF/beans.xml` is the easiest/quickest way to do that. I am not going to insist here. I did not "request changes" :)

Author: I didn't know very well the CDI spec (I'm more used to Spring than J2EE) that why I found strange to the `META-INF/beans.xml`. And there is always the risk that someone finding an empty file on a project just delete it by thinking it was unused ...

> or via the `quarkus.index-dependency` inside your `application.properties`.

Reviewer: ```suggestion
or via the `quarkus.index-dependency` property inside your `application.properties` (useful in cases where you can't modify the module).
```
